### PR TITLE
Fixes #9714: Remote run API should use relay API

### PR DESCRIPTION
**Pull request by VinceMacBuche**
https://www.rudder-project.org/redmine/issues/9714

### CODE REVIEW

File: rudder-web/src/main/scala/com/normation/rudder/web/rest/node/NodeAPIService8.scala [hunk 1]
```diff
@@ -138,54 +128,32 @@ class NodeApiService8 (
...
-        case None => // Nothing to do for now
-      }
-    }
-
-    // On exit will be called at the end of the process, or if any error/exception occurs, so we can close our outputstream here
-    override def onExit( statusCode : Int) = {
-      out.map(_.close())
-    }
+  def remoteRunRequest(nodeId: NodeId, classes : List[String], keepOutput : Boolean, asynchronous : Boolean) = {
+    val url = s"https://localhost/rudder/relay-api/remote-run/nodes/${nodeId.value}"
```
Comment (fanf) [inline rudder-web/src/main/scala/com/normation/rudder/web/rest/node/NodeAPIService8.scala]: This should be configured. I'm sure there will be case where relay-api won't be on the same machine than the web app (or behind an other root)

File: rudder-web/src/main/scala/com/normation/rudder/web/rest/node/NodeAPIService8.scala [hunk 2]
```diff
@@ -138,54 +128,32 @@ class NodeApiService8 (
...
-             case e : Throwable =>
-               out.close()
-               Failure(s"An error occured when applying policy on Node '${node.id.value}', cause is: ${e.getMessage}")
-           }
-    } yield {
-      runResponse(in)
-    }
+       httpResponse <- tryo { request.execute { runResponse } }
+       response <- if (httpResponse.isSuccess ) {
+         Full(httpResponse.body)
```
Comment (fanf) [inline rudder-web/src/main/scala/com/normation/rudder/web/rest/node/NodeAPIService8.scala]: I don't think this is streamed - is it?

Comment (VinceMacBuche) [inline rudder-web/src/main/scala/com/normation/rudder/web/rest/node/NodeAPIService8.scala]: request.execute allow to get data from the response as a stream ( inputStrem -> T ) we transform it using "runResponse" to an "OutputStream => Unit" which will read from the inputstream to put into the api output stream response and stream it


so yeah, body is not the "response data" but "how you want to get the response data" ... and it's quite misleading

Comment (fanf) [inline rudder-web/src/main/scala/com/normation/rudder/web/rest/node/NodeAPIService8.scala]: ok, better saw it without the diff view

**Comment by VinceMacBuche** [comment]
Commit modified

File: rudder-web/src/main/resources/configuration.properties.sample [hunk 3]
```diff
@@ -68,6 +68,10 @@ rudder.dir.lock=/var/rudder/lock/
 #
 rudder.endpoint.cmdb=http://localhost:8080/endpoint/upload/
 
+#
+# Location of the relay api used by rudder webapp
+#
+rudder.server.relay.api=localhost
```
Comment (fanf) [inline rudder-web/src/main/resources/configuration.properties.sample]: Here, we need to parametrize the whole address - nothing tell us what would be configured in multi-server in the future. 
I don't see any simple way to have: ```https://localhost/rudder/relay-api/remote-run/nodes/${nodeId}```, but we can use: ```https://localhost/rudder/relay-api/remote-run/nodes/%s" with a comment stating that "%s will be replaced by the node id".

File: rudder-web/src/main/scala/com/normation/rudder/web/rest/node/NodeAPIService8.scala [hunk 4]
```diff
@@ -138,54 +129,32 @@ class NodeApiService8 (
...
-        case None => // Nothing to do for now
-      }
-    }
-
-    // On exit will be called at the end of the process, or if any error/exception occurs, so we can close our outputstream here
-    override def onExit( statusCode : Int) = {
-      out.map(_.close())
-    }
+  def remoteRunRequest(nodeId: NodeId, classes : List[String], keepOutput : Boolean, asynchronous : Boolean) = {
+    val url = s"https://${relayApiEndpoint}/rudder/relay-api/remote-run/nodes/${nodeId.value}"
```
Comment (fanf) [inline rudder-web/src/main/scala/com/normation/rudder/web/rest/node/NodeAPIService8.scala]: and so here just use ```relayApiEndpoint.format(nodeId.value)```

File: rudder-web/src/main/resources/configuration.properties.sample (hunk 3 above)
Comment (VinceMacBuche) [inline rudder-web/src/main/resources/configuration.properties.sample]: Why not relay api endoit and just : https://localhost/rudder/relay-api

Comment (fanf) [inline rudder-web/src/main/resources/configuration.properties.sample]: Yes.

**Comment by VinceMacBuche** [comment]
Commit modified

**Pull request merged**